### ISSUE TITLE
exp: null-sequence contrastive guidance for contacts-v1 rollouts

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#321]
## Kind

- Kind: evals

## Question

Can a paired, null-sequence inference pass be used to subtract the sequence-independent contact-map prior during `contacts-v1` decoding, yielding more useful rollout diversity without sacrificing contact accuracy?

## Hypothesis

Yes, at moderate guidance strength. At decoding step `t`, run the same model on:

- the native sequence `x` followed by the generated contact-token prefix `y_<t`; and
- a same-length null sequence `x0` followed by the **identical** `y_<t`.

If their next-token logits are `l_x` and `l_0`, sample the native stream from

```text
g_gamma(v) = l_x(v) + gamma * (l_x(v) - l_0(v)).
```

`gamma=0` is ordinary decoding. Positive `gamma` amplifies tokens supported by the real sequence more than by the already-emitted contact map. This is CFG-like, but not literally classifier-free guidance: the second prompt is a sequence ablation, not a learned unconditional condition.

The raw ratio proposed in the motivating idea, `l_x - l_0`, is worth measuring as a diagnostic, but is not the safest primary decoder. A pure ratio can promote a token that is extremely unlikely under both streams merely because the null stream dislikes it slightly more. The primary implementation therefore retains the native logits as the base distribution and applies a native-probability plausibility gate before guidance.

Preregistered expectation:

- moderate guidance will increase **true-contact coverage across rollouts** and reduce rollout overlap while preserving consensus R-precision;
- very large `gamma`, pure-ratio decoding, or guidance on control tokens such as `<end>` will over-generate, increase malformed/low-probability output, or trade precision for indiscriminate novelty;
- the gain, if real, should survive both a homopolymer null and a composition-preserving shuffled-sequence null. A result unique to one null is a null-prompt artifact, not evidence for sequence-specific guidance.

## Background

- [#82](https://github.com/Open-Athena/MarinFold/issues/82) contains the strongest direct motivation. On three canonical proteins, replacing every residue with poly-alanine preserved most of the seeded-contact completion curve: with 32 true contacts in the prefix, native-sequence AUC was about 0.86-0.95 and poly-Ala AUC about 0.80-0.90. The model had learned a strong contact-map/topology prior and only a smaller sequence-dependent increment. The proposed decoder tries to isolate and amplify that increment online while forcing both passes to see exactly the same generated contacts.
- [#254](https://github.com/Open-Athena/MarinFold/issues/254) warns that raw diversity is not itself the bottleneck or the goal. One hundred ordinary rollouts already covered about 92% of true contacts on eval-val, and distinct one-contact seeds did not improve consensus. Success here must mean **precision-preserving useful diversity**, not merely more unique false pairs.
- [#306](https://github.com/Open-Athena/MarinFold/issues/306) and [#308](https://github.com/Open-Athena/MarinFold/issues/308) found that generic contact-block beam search and online novelty penalties worsened fold-mode coverage and/or ordinary accuracy at much higher cost. Sequence-contrastive guidance is different: it removes probability shared with a matched contact-history control instead of rewarding novelty indiscriminately.
- Use the current [#277](https://github.com/Open-Athena/MarinFold/issues/277) checkpoint, `contacts-v1-exp277-m2-p06-full-epoch-1.5B` at step 266344, so the result is comparable with the current rollout, fold-switching, beam, and novelty experiments.

### 1. Implement paired decoding with an exact shared suffix

Build one normal `contacts-v1` prompt realization per rollout. Construct the background prompt by copying its token IDs and replacing only amino-acid token IDs; keep document type, sequence length, position tokens, shuffled statement order, termini, and absolute token positions identical.

Decode the two streams with one model in a paired cached-forward loop. Only the native stream samples. Every sampled native token is then appended to **both** streams, so the background pass is always conditioned on the exact contact/control-token history that produced the main rollout; it never generates an independent background rollout.

Primary null: poly-Ala, because #82 already establishes it as a topology-only control. Sensitivity null: one deterministic composition-preserving permutation of the native amino acids per protein. Include poly-Lys only as a development diagnostic; do not select a final result from a large menu of homopolymers.

Primary guidance applies only when the decoder is choosing the two position tokens of a syntactically valid `<contact> <pI> <pJ>` statement. `<contact>` versus `<end>` and other control/grammar tokens use native logits, preventing contact-count changes from masquerading as better contact choice. An all-token-guidance arm is a development ablation.

Before applying guidance, restrict candidates to the native stream's ordinary top-p=0.95 set. Then combine logits and sample at T=1.0. This keeps a tiny null denominator from resurrecting tokens the native model considers implausible. Sweep `gamma in {0, 0.25, 0.5, 1, 2}` on development data. Also run the pure log-ratio score within the same native plausibility set as a labeled diagnostic, not as the assumed winner.

The worker must retain ordered per-rollout contact sets, completion/malformed status, generated-token count, and per-contact native/background log-probability differences. Add a unit test showing that the background suffix after `<begin_statements>` is token-for-token identical to the native suffix at every step.

### 2. Development and freeze

Use the 15 fold-switching development pairs from #304/#306 plus a fixed, length-stratified 16-protein subset of natural eval-val. Do not use eval-test.

First run a cheap teacher-forced diagnostic on ordinary-rollout prefixes: score completed contact statements under native and null prompts and ask whether the within-protein `log p_native - log p_null` separates true from false emitted contacts. Then run 20 rollouts per development protein for the guidance grid.

Freeze exactly one null, guidance scope, plausibility rule, and nonzero `gamma` before running the remaining natural eval-val proteins or the 29 primary fold-switching proteins. Selection criterion is the accuracy/diversity Pareto frontier, not diversity alone. If no nonzero setting preserves accuracy on development, stop and report the diagnostic rather than spending the full run.

### 3. Controls and full evaluation

Run 100 rollouts per protein for the frozen guided decoder and these controls:

1. the published iid recipe: T=1.0, top-p=0.95, no top-k, fresh document realization, token budget `6L+128`;
2. `gamma=0` in the new paired decoder, to isolate implementation/backend effects;
3. an ordinary temperature/top-p setting chosen on development to match the guided arm's rollout overlap or token entropy, to test whether guidance does more than flatten sampling.

Use common document realizations and sampling seeds across arms where semantics permit. Report both equal-rollout and equal-H100-second comparisons: guidance requires roughly two forward streams, so it must also compete with however many ordinary iid rollouts fit in the same measured time.

Full natural evaluation: all 97 eval-val proteins, with the 16 development proteins broken out and the other 81 reported separately. The zero-guidance control must reproduce #306's established exp277 iid result (R-precision about 0.5538 all-range and 0.5380 long-range) within 0.005 before guided results are interpreted.

Fold-diversity stress test: the 29 primary #304 fold-switching pairs, using the same fold-specific contact definitions and reference-blind 16-map selector as #306/#308. This cohort is no longer pristine to the project, so treat it as a fixed comparative stress test, seal predicted maps before scoring, and do not describe it as a newly unseen benchmark. A contact-map hit is not a recovered 3D fold without the existing structural validation gate.

### 4. Metrics

Compute curves at `N in {1, 2, 5, 10, 25, 50, 100}` from each 100-rollout pool.

Accuracy:

- standard vote-consensus R-precision, all-range and long-range;
- mean individual-rollout R-precision;
- completion rate, malformed rate, contacts per rollout, and vote concentration.

Useful diversity:

- true-contact union recall and number of distinct true contacts observed versus N;
- mean pairwise contact-set Jaccard similarity;
- number of distinct canonical contact maps and effective vote-support size;
- precision of contacts newly added to the union relative to iid.

Sequence-dependence diagnostics:

- within-protein AUC / paired difference for statement-level `log p_native - log p_null` on true versus false emitted contacts;
- guidance strength versus native likelihood, contact count, and malformed rate;
- agreement of effect direction between poly-Ala and composition-shuffled nulls.

Fold-switching:

- Fold1-only and Fold2-only recall, minority-mode enrichment, oracle dual-mode pool coverage, and reference-blind top-16 dual-mode coverage, using #304's definitions;
- equal-rollout and equal-time comparisons against iid, #306, and #308.

Use paired protein bootstrap intervals. Preserve per-input timing and worker metadata in `data/timings.csv` using the repository-wide predictor schema. Publish large rollout/logit artifacts to the public `open-athena/MarinFold` HF bucket and commit the small aggregate tables, plots, run manifest, and sealed-map hashes.

## Success criteria

1. **Validity gate:** `gamma=0` in the paired implementation reproduces the established exp277 iid eval-val all- and long-range R-precision within 0.005, with comparable completion and contact-count distributions. Otherwise stop and fix the decoder before interpreting guidance.
2. **Primary:** on the 81 non-development eval-val proteins, the frozen guided setting improves the area under the true-contact-union-recall-versus-log-N curve over both iid and the entropy-matched ordinary-sampling control, with a paired 95% bootstrap interval excluding zero, while all- and long-range consensus R-precision each regress by no more than 0.005.
3. **Compute check:** the useful-diversity advantage remains visible against the equal-H100-second iid control. If it exists only at equal rollout count, report it as a quality/compute tradeoff rather than an inference win.
4. **Secondary:** report whether guidance improves reference-blind or oracle dual-mode coverage on the fixed 29-pair fold-switching stress test. A fold-switching gain is not required for the general contact result, but a loss must be explicit.
5. Raw Jaccard reduction, unique-map count, or entropy alone does **not** count as success. If true-contact coverage or accuracy does not improve, the conclusion is that subtracting this null prior creates novelty rather than useful sequence-conditioned hypotheses.

## Notes

- This is inference-only: no model weights or training data change.
- A homopolymer is not a true unconditional prompt. The shuffled-composition sensitivity arm is required to distinguish a general sequence-contrastive effect from poly-Ala-specific out-of-distribution behavior.
- Keep eval-test unread.
- Implementation should initially live in the new experiment directory. Promote paired guidance into shared inference code only if the experiment is positive and a second use case exists.

**Comment by timodonnell** [comment]
🤖 Full experiment update: the paired decoder and frozen guidance arm have completed on all 97 eval-val proteins and all 29 primary fold-switch pairs. The gamma=0 implementation reproduces the published exp277 rollout score within the 0.005 validation gate.

After the experiment owner clarified that oracle best-of-100 matters more than consensus accuracy, I froze a validity-gated oracle endpoint (unfinished or malformed rollouts score zero) and rescored without changing the existing generations. On the 81 proteins not used for development, gamma=0.5 all-token guidance improves best-of-100 R-precision from 0.5331 to 0.5554 all-range (paired delta +0.02235, 95% bootstrap CI +0.0160 to +0.0289) and from 0.5279 to 0.5474 long-range (+0.01952, CI +0.0111 to +0.0281) versus ordinary iid sampling. It also beats the overlap-matched T=0.8 control.

A confirmatory pure native/null-ratio arm and a 200-rollout ordinary T=1.1 arm are now running at full scale. These were frozen before launch in `data/oracle_followup.json`; they test whether the gain is specific to bounded guidance and whether ordinary higher-temperature/equal-time sampling can match it. Final tables, public raw artifacts, README, and summary PDF will follow after those jobs complete.

**Comment by timodonnell** [comment]
🤖 Experiment complete and merged in #323.

**Primary result (81 eval-val proteins untouched during method selection):** bounded all-token poly-Ala guidance at gamma=0.5 improves validity-gated oracle best-of-100 R-precision from 0.5331 to 0.5554 all-range (paired delta +0.02235, 95% CI [+0.01588, +0.02899]) and from 0.5279 to 0.5474 long-range (+0.01952, CI [+0.01119, +0.02800]) versus ordinary iid sampling.

The gain survives equal-H100-time comparison: 100 guided rollouts match a median 132 iid rollouts and remain +0.01714 all / +0.01415 long, with both intervals excluding zero. The gamma=0 path reproduces the established exp277 score across all 97 eval-val proteins within 0.0007.

**Mechanism:** this is a per-sample quality gain, not increased diversity. Guidance raises mean single-rollout accuracy but increases pairwise Jaccard and lowers true-contact union recall. The literal native/null probability ratio is not useful: oracle accuracy falls to 0.4774/0.4713 and invalid/malformed output rises sharply. Keep the native logits and add a bounded contrast; do not sample from the pure ratio.

**Secondary fold stress test:** guidance is 6/29 oracle dual-mode and 3/29 blind top-16 versus iid 6/29 and 2/29. Pure ratio reaches 7/29 and 4/29, but with 5,329 malformed statements, so this is not a usable win.

Artifacts:
- [README and full result tables](https://github.com/Open-Athena/MarinFold/tree/main/experiments/exp321_evals_null_sequence_contrastive_guidance_for_contacts)
- [summary PDF](https://github.com/Open-Athena/MarinFold/blob/main/experiments/exp321_evals_null_sequence_contrastive_guidance_for_contacts/plots/summary.pdf)
- [public raw rollout/timing parquets](https://huggingface.co/buckets/open-athena/MarinFold/tree/main/data/exp321/null-sequence-guidance-v1) — 2,318 files, anonymously verified
- [merged PR #323](https://github.com/Open-Athena/MarinFold/pull/323)

All 120 full jobs succeeded; completeness, timing-schema, sealed-map, compile, unit-test, and public-artifact checks pass. The issue remains open for human disposition, per experiment policy.
